### PR TITLE
Fixes #20513 - shareable locale Makefile and Rakefile definitions

### PR DESCRIPTION
**Pull request by tstrachota**
This change will allow plugins to share common parts of rake tasks and make targets around extracting strings for translation.

### CODE REVIEW

**Comment by tstrachota** [comment]
@mbacovsky could you please to the review?

File: doc/i18n.md [hunk 1]
```diff
@@ -62,6 +61,29 @@ locale
...
+VERSION = $(shell ruby -e 'require "rubygems"; spec = Gem::Specification::load("../hammer_cli_awesome.gemspec"); puts spec.version')
+MAIN_MAKEFILE = $(shell ruby -e 'require "hammer_cli"; puts HammerCLI::I18n.main_makefile')
+
+include $(MAIN_MAKEFILE)
+```
+
+Make sure you have a project created in [transifex](www.transifex.com) and correct configuration stored in your plugins repository ([example config file](../.tx/config)).
+
+
+Calling `make -C ./locale tx-update` will then extract new strings, update `.po` and `.mo` files and commit the changes.
```
[the remark this answers is not in the document]
Comment (tstrachota) [reply doc/i18n.md]: I added a sentence that warns developers. Using `bundle exec` is imho correct and standard way. What was your idea of finding the main makefile location without using bundler?

**Comment by tstrachota** [comment]
Updated

**Comment by tstrachota** [comment]
[test]

File: doc/i18n.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbacovsky) [reply doc/i18n.md]: The point was it was not necessary in the past to use bundler to invoke the make. From the users perspective the make suddenly fails with backtrace if used as usual. I'd expect Makefile to handle the situation and use bundler as a fallback option on its own (as `make` usually does in other projects).

File: locale/Makefile [hunk 2]
```diff
@@ -7,60 +7,7 @@
...
-VERSION = $(shell ruby -e 'require "rubygems";spec = Gem::Specification::load("../hammer_cli.gemspec");puts spec.version')
-POTFILE = $(DOMAIN).pot
-MOFILE = $(DOMAIN).mo
-POFILES = $(shell find . -name '$(DOMAIN).po')
-MOFILES = $(patsubst %.po,%.mo,$(POFILES))
-POXFILES = $(patsubst %.po,%.pox,$(POFILES))
-EDITFILES = $(patsubst %.po,%.edit.po,$(POFILES))
-TIMESTAMPFILES = $(patsubst %.po,%.po.time_stamp,$(POFILES))
+VERSION = $(shell ruby -e 'require "rubygems"; spec = Gem::Specification::load("../hammer_cli.gemspec"); puts spec.version')
+MAIN_MAKEFILE = $(shell ruby -e 'require "hammer_cli"; puts HammerCLI::I18n.main_makefile')
```
Comment (mbacovsky) [inline locale/Makefile]: @tstrachota I found out we use `bundle exec` already in the `extract-strings` target so I think it makes sense to include it here as well. It is pointless to expect calling any of the targets outside of devel env anyway, isn't it?

File: locale/Makefile.def [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#
+
+POTFILE = $(DOMAIN).pot
+MOFILE = $(DOMAIN).mo
+POFILES = $(shell find . -name '$(DOMAIN).po')
+MOFILES = $(patsubst %.po,%.mo,$(POFILES))
+POXFILES = $(patsubst %.po,%.pox,$(POFILES))
+EDITFILES = $(patsubst %.po,%.edit.po,$(POFILES))
+TIMESTAMPFILES = $(patsubst %.po,%.po.time_stamp,$(POFILES))
+
```
Comment (mbacovsky) [inline locale/Makefile.def]: How about adding help target?
```bash
.PHONY: help

help: ## This help message
	@echo -e "$$(grep -hE '^\S+:.*##' $(MAKEFILE_LIST) | sed -e 's/:.*##\s*/:/' -e 's/^\(.\+\):\(.*\)/\\x1b[36m\1\\x1b[m:\2/' | column -c2 -t -s :)"
```
Then commenting the target with double hash like: `clean: ## Clean everything` would produce:

```
$ make help
help   This help message
...
check  Check translations using translate-tool
clean  Clean everything
```
Seems useful. I've seen it adopted by some other projects

File: locale/Makefile.def [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+%.mo: %.po
+	mkdir -p $(shell dirname $@)/LC_MESSAGES
+	msgfmt -o $(shell dirname $@)/LC_MESSAGES/$(MOFILE) $<
+
+# Generate MO files from PO files
+all-mo: $(MOFILES)
```
Comment (mbacovsky) [inline locale/Makefile.def]: As this is going to be shared and we no longer don't have control over what is in the project dirs I'd suggest to add .PHONY for non-file targets. Note: Makefile style guide suggests to add `.PHONY: ` immediately before the target rather then listing all at one place.
```
.PHONY: all-mo
```

File: locale/Makefile.def (hunk 3 above)
Comment (tstrachota) [reply locale/Makefile.def]: Nice, I added that.

**Comment by tstrachota** [comment]
@mbacovsky updated. I added `bundle exec` to the commands. It's true that we're already using it in `extract-strings` target.

File: locale/Makefile.def [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+extract-strings:
+	bundle exec rake gettext:find
+
+.PHONY: tx-update
+tx-update: extract-strings tx-pull update-po $(MOFILES) ## Download and merge translations from Transifex
+	git add ../locale
+	git commit -m "i18n - extracting new, pulling from tx"
+	@echo Changes commited!
+
+.PHONY:clean
```
Comment (mbacovsky) [inline locale/Makefile.def]: nitpick - space missing

File: locale/Makefile.def [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+# Merge .edit.po into .po
+.PHONY: update-po
+update-po:
+	for f in $(EDITFILES); do \
+		msgcat --use-first --no-location $$f $${f//.edit/} > $${f//.edit/}; \
+	done
+
+# Extract strings and update the .pot, prepare .edit.po files
+.PHONY: extract-strings
+extract-strings:
```
Comment (mbacovsky) [inline locale/Makefile.def]: @tstrachota, were some of the target not documented/added to the help on purpose?

Comment (tstrachota) [reply locale/Makefile.def]: Yes, I intentionally documented only the main targets to keep the help simple and left out targets that are rather "intermediate".

**Comment by tstrachota** [comment]
Updated

**Comment by mbacovsky** [comment]
Thanks, @tstrachota :banana: :gorilla: 
:+1:

**Pull request merged**